test/extensions/filters/http/cache/cache_filter_integration_test.cc

```diff
@@ -297,6 +297,348 @@ TEST_P(CacheIntegrationTest, GetRequestWithBodyAndTrailers) {
+    waitForNextUpstreamRequest();
+    upstream_request_->encodeHeaders(response_headers, /*end_stream=*/true);
+    response_decoder->waitForEndStream();
+    EXPECT_TRUE(response_decoder->complete());
+    EXPECT_THAT(response_decoder->headers(), IsSupersetOfHeaders(response_headers));
+    EXPECT_TRUE(response_decoder->headers().get(Http::Headers::get().Age).empty());
+    EXPECT_EQ(response_decoder->body().size(), 0);
+    EXPECT_THAT(waitForAccessLog(access_log_name_), testing::HasSubstr("- via_upstream"));
+  }
+
+  // Advance time, to verify the original date header is preserved.
+  simTime().advanceTimeWait(Seconds(10));
+
+  // Send second request, and get response from upstream, since the head requests are not stored in
+  // cache.
+  {
+    IntegrationStreamDecoderPtr response_decoder =
+        codec_client_->makeHeaderOnlyRequest(request_headers);
+    waitForNextUpstreamRequest();
+    upstream_request_->encodeHeaders(response_headers, /*end_stream=*/true);
+    response_decoder->waitForEndStream();
+    EXPECT_TRUE(response_decoder->complete());
+    EXPECT_THAT(response_decoder->headers(), IsSupersetOfHeaders(response_headers));
+    EXPECT_EQ(response_decoder->body().size(), 0);
+    EXPECT_TRUE(response_decoder->headers().get(Http::Headers::get().Age).empty());
+    // Advance time to force a log flush.
+    simTime().advanceTimeWait(Seconds(1));
+    EXPECT_THAT(waitForAccessLog(access_log_name_), testing::HasSubstr("- via_upstream"));
+  }
+}
+
+TEST_P(CacheIntegrationTest, ServeHeadFromCacheAfterGetRequest) {
+  useAccessLog("%RESPONSE_FLAGS% %RESPONSE_CODE_DETAILS%");
+  simTime().setSystemTime(std::chrono::hours(1));
+  initializeFilter(default_config);
+
+  const Http::TestRequestHeaderMapImpl get_request_headers = {
+      {":method", "GET"},
+      {":path", absl::StrCat("/", protocolTestParamsToString({GetParam(), 0}))},
+      {":scheme", "http"},
+      {":authority", "ServeHeadFromCacheAfterGetRequest"}};
+
+  const std::string response_body(42, 'a');
+  Http::TestResponseHeaderMapImpl response_headers = {
+      {":status", "200"},
+      {"date", formatter_.now(simTime())},
+      {"cache-control", "public,max-age=3600"},
+      {"content-length", std::to_string(response_body.size())}};
+
+  // Send GET request, and get response from upstream.
+  {
+    IntegrationStreamDecoderPtr response_decoder =
+        codec_client_->makeHeaderOnlyRequest(get_request_headers);
+    waitForNextUpstreamRequest();
+    upstream_request_->encodeHeaders(response_headers, /*end_stream=*/false);
+    upstream_request_->encodeData(response_body, /*end_stream=*/true);
+    // Wait for the response to be read by the codec client.
+    response_decoder->waitForEndStream();
+    EXPECT_TRUE(response_decoder->complete());
+    EXPECT_THAT(response_decoder->headers(), IsSupersetOfHeaders(response_headers));
+    EXPECT_TRUE(response_decoder->headers().get(Http::Headers::get().Age).empty());
```

Comment:
minor suggestion: this is all over this file in pre-existing code, but if you compare a string to "" with EXPECT_EQ, you'll get more potentially actionable info in the test log if it fails.

Feel free to either ignore this or fix all of them :)

Reply:
Thank you for your suggestion. I tried following it, however, turns out that `response_decoder->headers().get(Http::Headers::get().Age)` returns an `Envoy::Http::HeaderMap::GetResult`,  which contains an `absl::InlinedVector<HeaderEntry*, 1>`. Therefore, as far as I understand, empty() is to check for an empty list and not an empty string.

Reply:
Ah in that case compare size to 0, or better yet compare the vector to the expected empty vector using the gtest macros for that.